Add FileUtils::VERSION

junaruga opened the pull request:
Hello,
I want to add `FileUtils::VERSION`.

Because `bundler` is bundling `fileutils` internally. And I want to know the file's original version casually.
https://github.com/bundler/bundler/blob/master/lib/bundler/vendor/fileutils/lib/fileutils.rb

I tested below things on latest trunk's Ruby.

```
$ ruby -v
ruby 2.5.0dev (2017-11-09 trunk 60727) [x86_64-linux]

$ ruby -Ilib -r fileutils -e 'puts FileUtils::VERSION'
0.7.3-dev

$ gem --version
2.7.2

$ which gem
/home/jaruga/git/ruby/dest/bin/gem

$ gem build fileutils.gemspec
WARNING:  description and summary are identical
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: fileutils
  Version: 0.7.3.pre.dev
  File: fileutils-0.7.3.pre.dev.gem
```

I referred `did_you_mean` for the implementation of `FileUtils::VERSION`.
https://github.com/yuki24/did_you_mean/blob/master/did_you_mean.gemspec

hsbt commented on lib/fileutils/version.rb:
```diff
@@ -0,0 +1,3 @@
+module FileUtils
+  VERSION = "0.7.3-dev"
```
Can you update this version to "0.7.3.beta1" ?

junaruga replied:
Sure, changed.

amatsuda commented:
You can get the version via `Gem.loaded_specs` when fleutils is loaded as a gem.

```
% ruby -rfileutils -e 'puts Gem.loaded_specs["fileutils"]'
#<Gem::Specification name=fileutils version=0.7.2>
```

Wouldn't this be enough for you?

junaruga commented:
@hsbt ok, I changed the version name! and rebased.

Now the result is like this.

```
$ gem build fileutils.gemspec
WARNING:  description and summary are identical
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: fileutils
  Version: 0.7.3.beta1
```

@amatsuda thanks for the information. I did not know the way.
However when I ran the command right now, the version is still `0.7.2`.

```
$ ruby -rfileutils -e 'puts Gem.loaded_specs["fileutils"]'
#<Gem::Specification name=fileutils version=0.7.2>
```

I suppose https://github.com/ruby/ruby/blob/trunk/lib/fileutils.gemspec is loaded from above command.

amatsuda commented:
> However when I ran the command right now, the version is still 0.7.2.

Nah, that is totally unrelated to how the gem internally sets the version number. If `Gem.loaded_specs["fileutils"]` returns 0.7.2, that just means you loaded 0.7.2 gem.

I could actually do this locally using bundler
```
% bundle e ruby -e 'puts Gem.loaded_specs["fileutils"]'
#<Gem::Specification name=fileutils version=0.7.3.beta1>
```
simply with this change for the fileutils codebase.

```
diff --git a/fileutils.gemspec b/fileutils.gemspec
index b50bbf3..c1feeb5 100644
--- a/fileutils.gemspec
+++ b/fileutils.gemspec
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 Gem::Specification.new do |s|
   s.name = "fileutils"
-  s.version = '0.7.2'
+  s.version = '0.7.3.beta1'
   s.date = '2017-02-06'
```

So, I don't think we need to merge your proposed change to fulfill your use case.

junaruga commented:
@amatsuda sorry. Now I got the same result with you on your condition, just changing the `s.version`.
However I got an error on my proposed modification.

```
$ ruby -v
ruby 2.5.0dev (2017-11-09 trunk 60727) [x86_64-linux]

$ bundle -v
Bundler version 1.16.0

$ git branch
* feature/add-version-var
  master

$ bundle install --path vendor/bundle
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Fetching rake 12.2.1
Installing rake 12.2.1
Using bundler 1.16.0
Using fileutils 0.7.3.beta1 from source at `.`
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Bundled gems are installed into `./vendor/bundle`

$ bundle list
Gems included by the bundle:
  * bundler (1.16.0)
  * fileutils (0.7.3.beta1)
  * rake (12.2.1)

$ bundle e ruby -e 'puts Gem.loaded_specs["fileutils"]'
Traceback (most recent call last):
	9: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	8: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	7: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler/setup.rb:10:in `<top (required)>'
	6: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler.rb:107:in `setup'
	5: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler/runtime.rb:26:in `setup'
	4: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler/runtime.rb:26:in `map'
	3: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/forwardable.rb:229:in `each'
	2: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/forwardable.rb:229:in `each'
	1: from /home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler/runtime.rb:31:in `block in setup'
/home/jaruga/git/ruby/dest/lib/ruby/2.5.0/bundler/runtime.rb:313:in `check_for_activated_spec!': You have already activated fileutils 0.7.2, but your Gemfile requires fileutils 0.7.3.beta1. Since fileutils is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports fileutils as a default gem. (Gem::LoadError)
```

> So, I don't think we need to merge your proposed change to fulfill your use case.

I still do not understand why you think the proposed change is not needed.

My use case is

Bundler includes modified `fileutils` without `fileutils.gemspec`.

https://github.com/bundler/bundler/blob/master/lib/bundler/vendor/fileutils/lib/fileutils.rb
> module Bundler::FileUtils

The module name is changed from the original one.

`fileutils.rb` does not have `VERSION.`

Other bundled libraries have `VERSION`. This is helpful for me.

> https://github.com/bundler/bundler/blob/master/lib/bundler/vendor/net-http-persistent/lib/net/http/persistent.rb#L206
> ```
>   VERSION = '2.9.4'
> ```

junaruga commented:
I have no idea about how to fix this. Could you give me an advice?
